### PR TITLE
Unify the FV3-HRRR and FV3-HRRR-smoke

### PR DESCRIPTION
**Pull request by haiqinli**
1). The codes of ufs-weather-model/fv3atm/ccpp-physics is based on RRFS_dev for baseE.

2). The FV3_HRRR and FV3_HRRR_smoke are unified as one suite of FV3_HRRR. If the input namelist option rrfs_sd is false, it runs as the standard alone RRFS. IF rrfs_sd is true, it runs with smoke and dust for RRFS-SD. The meteorological fields are identical between the runs with rrfs_sd=T or rrfs_sd=F.

3). There is just one field_table, which contains smoke and dust tracers. We use the rrfs_sd option to control the tracers index. If rrfs_sd is false, only the default tracers of RRFS are read in. If rrfs_sd is true, the smoke and dust tracers for RRFS-SD are added.

4). The dry deposition velocity is an input variable to MYNN PBL.

5). The smoke and dust codes are significantly cleaned and improved.

### CODE REVIEW

**Comment by hu5970** [comment]
Please review this PR at your earliest convenience. This PR will be a part of RRFS baseline E. Thanks, Ming

**Review by zhanglikate — APPROVED** [review]

File: physics/module_bl_mynn.F90 [hunk 1]
```diff
@@ -1066,7 +1066,7 @@ SUBROUTINE mynn_bl_driver(            &
                 enddo
              ELSE
                 do ic = 1,ndvel
-                   vd1(ic) = 0. !is this correct??? (ite) or (ndvel)
+                   vd1(ic) = 0. ! dry deposition veolocity
```
Comment (SamuelTrahanNOAA) [inline physics/module_bl_mynn.F90]:
```suggestion
                   vd1(ic) = 0. ! dry deposition velocity
```

File: physics/module_bl_mynn.F90 [hunk 2]
```diff
@@ -1052,9 +1052,9 @@ SUBROUTINE mynn_bl_driver(            &
           ENDDO ! end k
 
           !initialize smoke/chem arrays (if used):
-             IF  ( rrfs_smoke .and. mix_chem ) then
+             IF  ( rrfs_sd .and. mix_chem ) then
                 do ic = 1,ndvel
-                   vd1(ic) = vdep(i,ic) !is this correct????
+                   vd1(ic) = vdep(i,ic) ! dry deposition veolocity
```
Comment (SamuelTrahanNOAA) [inline physics/module_bl_mynn.F90]:
```suggestion
                   vd1(ic) = vdep(i,ic) ! dry deposition velocity
```

**Review by SamuelTrahanNOAA — APPROVED** [review]

File: physics/module_bl_mynn.F90 (hunk 2 above)
Comment (haiqinli) [reply physics/module_bl_mynn.F90]: Thanks, Sam. It was corrected accordingly.

**Comment by hu5970** [comment]
@haiqinli Please merge this one if it is ready. Thanks, Ming

**Comment by SamuelTrahanNOAA** [comment]
> @haiqinli Please merge this one if it is ready. Thanks, Ming

We need pull requests at the fv3atm and ufs-weather-model level first, otherwise merging to ccpp-physics would accomplish nothing.

Also, we need to run regression tests. Given the urgency of your message, I can run them myself instead of waiting for Jacques. (I'm tasked for emergency support of the parallels).

Hera is down tomorrow, so we need to begin testing today.

**Comment by SamuelTrahanNOAA** [comment]
The higher-level PRs are here, so I will begin regression tests now:

https://github.com/NOAA-GSL/ufs-weather-model/pull/148
https://github.com/NOAA-GSL/fv3atm/pull/154

**Comment by SamuelTrahanNOAA** [comment]
This branch does not work; several tests fail to compile or crash. See the ufs-weather-model PR for details:

https://github.com/NOAA-GSL/ufs-weather-model/pull/148

File: physics/module_bl_mynn.F90 [hunk 3]
```diff
@@ -5314,15 +5314,16 @@ SUBROUTINE mynn_mix_chem(kts,kte,i,     &
        khdz(k) = MAX(khdz(k), -0.5*(s_aw(k)-s_aw(k+1)))
     ENDDO
 
-    !Enhance diffusion over fires
-    IF ( fire_turb ) THEN
+    !Enhanced mixing over fires
+    IF ( rrfs_sd .and. enh_mix ) THEN
+       kmaxfire = ceiling(log(frp))
```
Comment (SamuelTrahanNOAA) [inline physics/module_bl_mynn.F90]:
```suggestion
```

File: physics/module_bl_mynn.F90 [hunk 4]
```diff
@@ -5314,15 +5314,16 @@ SUBROUTINE mynn_mix_chem(kts,kte,i,     &
...
        DO k=kts+1,kte-1
           khdz_old  = khdz(k)
           khdz_back = pblh * 0.15 / dz(k)
           !Modify based on anthropogenic emissions of NO and FRP
           IF ( pblh < pblh_threshold ) THEN
-             IF ( emis_ant_no > no_threshold ) THEN
-                khdz(k) = MAX(1.1*khdz(k),sqrt((emis_ant_no / no_threshold)) / dz(k) * rhoz(k)) ! JLS 12/21/21
+             IF ( emis_ant_no > NO_threshold ) THEN
+                khdz(k) = MAX(1.1*khdz(k),sqrt((emis_ant_no / NO_threshold)) / dz(k) * rhoz(k)) ! JLS 12/21/21
```
Comment (SamuelTrahanNOAA) [inline physics/module_bl_mynn.F90:5326]:
```suggestion
                kmaxfire = ceiling(log(frp))
                khdz(k) = MAX(1.1*khdz(k),sqrt((emis_ant_no / NO_threshold)) / dz(k) * rhoz(k)) ! JLS 12/21/21
```

**Review by SamuelTrahanNOAA — CHANGES_REQUESTED** [review]
When frp is 0, code fails with floating point exception.

File: physics/module_bl_mynn.F90 (hunk 3 above)
Comment (SamuelTrahanNOAA) [reply physics/module_bl_mynn.F90]: Also, a few lines down, you need to insert kmaxfire here:

```
             IF ( frp > frp_threshold ) THEN
                kmaxfire = ceiling(log(frp))
                khdz(k) = MAX(1.1*khdz(k),sqrt((emis_ant_no / NO_threshold)) / dz(k) * rhoz(k)) ! JLS 12/21/21
```

**Comment by SamuelTrahanNOAA** [comment]
I've found the bugs that cause RRFS to crash early in debug mode. Thanks to that, I was able to run the smoke code in debug mode, and found one bug:

```
             kmaxfire = ceiling(log(frp))
```

when frp=0, this results in a floating point exception. The calculation must be moved to within the "if" block:

```
             IF ( frp > frp_threshold ) THEN
                kmaxfire = ceiling(log(frp))
                khdz(k) = MAX(1.1*khdz(k), (1. - k/(kmaxfire*2.)) * ((log(frp))**2.- 2.*log(frp)) / dz(k)*rhoz(k)) ! JLS 12/21/21
```

I'm working on a PR to the authoritative to fix the other crashes. The bugs are all in fv_regional_bc.F90

**Comment by SamuelTrahanNOAA** [comment]
@haiqinli I'm testing some changes to the module_sf_ruclsm bug fixes. These changes were requested by @tanyasmirnova because the bug fixes belong in lsm_ruc.F90 and should also include "snow on ice" sanity checks. I'm doing regression tests now. I don't expect the results to change, but it is possible if a point has an unphysically small amount of snow on ice (<1e-20)

**Review by tanyasmirnova — APPROVED** [review]

**Review by tanyasmirnova — APPROVED** [review]

**Review by SamuelTrahanNOAA — APPROVED** [review]

**Pull request merged**